addons/crm/models/crm_lead.py

```diff
@@ -892,6 +900,76 @@ def redirect_lead_opportunity_view(self):
             'context': {'default_type': self.type}
         }
 
+    def smart_meeting_calendar_view(self, opportunity_id):
+        meeting_results = self.env["calendar.event"].search_read([('opportunity_id', 'in', [opportunity_id])], ['start', 'stop', 'allday'])
+        if not meeting_results:
+            return "week", False
+
+        # User TZ and Now
+        user_tz = self.env.context.get('tz')
+        pytz_user_tz = pytz.timezone(user_tz)
+        now_utc = datetime.utcnow()
+        now_user_naive = datetime.now().astimezone(pytz_user_tz).replace(tzinfo=None)
+
+        # Data and declaration
+        start_dates = [row.get('start') for row in meeting_results]
+        stop_dates = [row.get('stop') for row in meeting_results]
+        is_allday = [row.get('allday') for row in meeting_results]
+        n_meetings = len(start_dates)
+        past_meetings_indices, future_meetings_indices = [], []
+
+        # Get indices of relevant meetings: future ones if any, past otherwise.
```

Comment:
I have to admit I did not understand everything in this code. According to commit message, you try to do
```
-If no meetings: show current week
-If one: show that week
-If more than one and same week : show that week
-If more than one and same month : show that month
-If more than one month, show month of the first meeting.
```
First two cases are straightforward and could lead to an early return (already done for first case aka no result).

Other cases computation seem complicated. I understand you have to play with "all day events" / "time limited events" as well as TZ. Then there is this past / future thingy that is unclear, as well as index management that lead to code hard to read. Especially that this method lacks comments.

Naively I would say that you could try something like
  * fetch meetings
  * compute a start / end in user timezone for all events, either moving UTC-stored when having start and stop, either setting from 0 to 23:59 if missing information. Store it into a tuple or a light structure so that we know always work in user TZ;
  * take subset of not yet finished events if any, or all events if finished;
  * compare week of first and last events and return accordingly (using isocalendar you can have week number directly, easing comparison);

It seems to be globally what you do. It is just a matter of trying to have clear steps in computation and clear computation of what we expect and how we classify records.

Last but not least: tests :D .

Thanks !